### PR TITLE
Cleans up key handling in drivers (was "fixes VkeyPacketSimulator is broken")

### LINKED ISSUE #3054: `VkeyPacketSimulator` scenario is broken
_Originally posted by @BDisp in https://github.com/gui-cs/Terminal.Gui/issues/2927#issuecomment-1857785921_

It's been broken in `v2_develop` for a while. It is made both better and worse in #2927.

This issue is to track fixing it after #2927 is merged.

### PR DESCRIPTION
**Pull request by BDisp**
Fixes a lot of keyboard related things, but does NOT fix #3054 

## Pull Request checklist:

- [c] I've named my PR in the form of "Fixes #issue. Terse description."
- [c] My code follows the [style guidelines of Terminal.Gui](https://github.com/gui-cs/Terminal.Gui/blob/develop/.editorconfig) - if you use Visual Studio, hit `CTRL-K-D` to automatically reformat your files before committing.
- [x] My code follows the [Terminal.Gui library design guidelines](https://github.com/gui-cs/Terminal.Gui/blob/develop/CONTRIBUTING.md)
- [x] I ran `dotnet test` before commit
- [x] I have made corresponding changes to the API documentation (using `///` style comments)
- [x] My changes generate no new warnings
- [x] I have checked my code and corrected any poor grammar or misspellings
- [x] I conducted basic QA to assure all features are working

### CODE REVIEW

**Review by dodexahedron — CHANGES_REQUESTED** [review]

File: UnitTests/Input/KeyTests.cs [hunk 1]
```diff
@@ -337,4 +329,15 @@ public void TryParse_ShouldReturnTrue_WhenValidKey (string keyString, Key expect
...
 	[InlineData ("Ctrl-Ctrl")]
 	public void TryParse_ShouldReturnFalse_On_InvalidKey (string keyString) => Assert.False (Key.TryParse (keyString, out var _));
+
+	[Theory]
+	[InlineData (KeyCode.A, KeyCode.A)]
+	[InlineData (KeyCode.F1, KeyCode.F1)]
+	public void Casting_Between_Key_And_KeyCode (KeyCode keyCode, KeyCode key)
+	{
+		Assert.Equal (keyCode, (Key)key);
+		Assert.NotEqual (keyCode, ((Key)key).WithShift);
```
[the remark this answers is not in the document]
Comment (BDisp) [reply UnitTests/Input/KeyTests.cs:339]: But KeyCode.A is equal to Key.A which represents the lower case without shift. So both are lower case. I think it's correct.

[the remark this answers is not in the document]
Comment (BDisp) [reply UnitTests/Input/KeyTests.cs:339]: Before the `Key.A` (now `KeyCode`) was considered as a capital letter `'A'` because the `Key.ShiftMask` was omitted. Now it's accomplish as `KeyCode.A | KeyCode.ShiftMask` to represent the capital `A` character. I think this change confuse you.

[the remark this answers is not in the document]
Comment (dodexahedron) [reply UnitTests/Input/KeyTests.cs:339]: Quite possibly. And it may also only matter on Windows.

[the remark this answers is not in the document]
Comment (BDisp) [reply UnitTests/Input/KeyTests.cs:339]: Yes, on CursesDriver and NetDriver we have to map the modifiers because they only provide character and ANSI escape, to be consistent on all drivers.

[the remark this answers is not in the document]
Comment (BDisp) [reply UnitTests/Input/KeyTests.cs:339]: `ConsoleKeyInfo` only handles the `shift`, `alt` and `control` because it use `ConsoleModifiers` which support only theses three modifiers. So the `capslock`  only can be handles by the `WindowsDriver` which use the `ConsoleKeyInfoEx` that store more the capslock, numlock and scrolllock. For this reason there is no sense to make unit tests that handles with this extensions modifiers because `dotnet` doesn't provide library for them, only Win32 API.

**Comment by BDisp** [comment]
@tig have you at least looked at this PR to understand the reason for the changes?

**Comment by tig** [comment]
No, I've not had a chance. But I will asap.

**Comment by tig** [comment]
What is the setup required to actually get the VK_PACKET stuff
to come into play?

Is it an Telnet/ssh thing. ?

**Comment by BDisp** [comment]
> What is the setup required to actually get the VK_PACKET stuff to come into play?
> 
> Is it an Telnet/ssh thing. ?

Yes,  know it's used with RDP (Remote Desktop Protocol) where the unicode char is encoded with the virtual key.

**Comment by BDisp** [comment]
Typing in the first `TextView` mirrors all on the second `TextView` on the VK Packet Simulator.

**Comment by tig** [comment]
> > What is the setup required to actually get the VK_PACKET stuff to come into play?
> > Is it an Telnet/ssh thing. ?
> 
> Yes, know it's used with RDP (Remote Desktop Protocol) where the unicode char is encoded with the virtual key.

I understand what the simulator is supposed to do.

I want to be able to use VK Packet support "for real". Do you know what I need to do in order to do that?

**Comment by BDisp** [comment]
> > > What is the setup required to actually get the VK_PACKET stuff to come into play?
> > > Is it an Telnet/ssh thing. ?
> > 
> > 
> > Yes, know it's used with RDP (Remote Desktop Protocol) where the unicode char is encoded with the virtual key.
> 
> I understand what the simulator is supposed to do.
> 
> I want to be able to use VK Packet support "for real". Do you know what I need to do in order to do that?

I never try it with a real support but here (https://github.com/gui-cs/Terminal.Gui/issues/2008#issuecomment-1246460048) is a way to try it.

**Comment by tig** [comment]
@BDisp, you wrote this:

image

You are not wrong in saying "'â' is the lower case of 'Â`". However, in all of my testing, on the US ENG, DEU, POR, and other keyboards this is NOT how keyboards work, or what users expect.

For example, in either Windows Terminal settings, use the settings UI to change the keybinding for "New tab, proifle index 0"

image

Using the DEU keyboard, the "CTRL+SHIFT+ä" key (which in ENG is "CTRL+SHIFT+'" gets displayed as "Ctrl+Shift+ä" NOT "Ctrl-Shift+Ä"

The same is true for, eg.., the Key.D7.

**Comment by BDisp** [comment]
> @BDisp, you wrote this:
> 
> image
> You are not wrong in saying "'â' is the lower case of 'Â`". However, in all of my testing, on the US ENG, DEU, POR, and other keyboards this is NOT how keyboards work, or what users expect.
> 
> For example, in either Windows Terminal settings, use the settings UI to change the keybinding for "New tab, proifle index 0"
> 
> image
> Using the DEU keyboard, the "CTRL+SHIFT+ä" key (which in ENG is "CTRL+SHIFT+'" gets displayed as "Ctrl+Shift+ä" NOT "Ctrl-Shift+Ä"
> 
> The same is true for, eg.., the Key.D7.

@tig I have a Portuguese keyboard and I know what I'm talking about. To get the `â` I need type `shift+^+A` but the shift modifier doesn't matter because it's only to get the `^`. After type the `A` the result is `â` (remember that the `A` without shift represent the lower `a`).

image

 To get the `Â` I need to type `shift+^+shift+A` , but the shift modifier doesn't matter because it's only to get the `^` but the second shift modifier to get the upper case `A` matter. After type `A` the result is `Â`. 

image

Using a virtual keyboard isn't the same as a real keyboard. So a `Ctrl+Shift+â` is never possible because other modifier is been used, the Ctrl key. Only is possible `Ctrl+Shift+A`.
If I type `Ctrl+Shift+/` I get this on `WindowsDriver`:

image

If I type `Shift+/` I get the follow and we could consider as `KeyCode.ShiftMask | (KeyCode)'/'`

image

**Comment by BDisp** [comment]
My keyboard has more or less this layout:

image

Pressing `Ctrl+Shift+~` in WT the result is `ctrl+shift+~`. Although I have the shift key pressed it's ignored if another modifier key, like the Ctrl or Alt, are also pressed. So only the lower character is considered (`~`).

image

Pressing `Shift+~` in WT the result is `shift+~`. WT only consider the lower character (`~`) which isn't the right binding.

image

If I press `Shift+~` in the Windows notepad I get nothing and only when I press `Shift+~` again I get two `^^`. This is normally for accented characters because the terminal is waiting for the next key pressed to write the right one. If I press `Shift+~` and release and then press `a` I get `â`.  If I press `Shift+~` and release and then press `Shift+a` I get `Â`. This last normally we don't need to release the keys and is done at once, like `Shift+~+a` or `Shift+~` and without release the shift press `a`.

image

Pressing `Ctrl+Shift+7` in WT the result is `ctrl+shift+7`. Although I have the shift key pressed it's ignored if another modifier key, like the Ctrl or Alt, are also pressed. So only the lower character is considered (`7`).

image

Pressing `Shift+7` in WT the result is `shift+7`. WT only consider the lower character (`7`) which isn't the right binding.

image

If I press `Shift+7` in the Windows notepad I get `/`. This is right behavior. So, don't rely on the WT to test the right key bindings because we need keys that interpret the right writings as well. It's better to use the notepad to test the printable keys.

image

**Comment by tig** [comment]
@BDisp please read this: https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-mapvirtualkeya

Esp:

image

This is the mode I believe Terminal.Gui should operate in. 

The current code in this PR (and v2_develop) is broken w.r.t. keys like OemPlus and more. Neither of you nor I had this correct.

The `_scanCodes` HashSet is very broken (and has been since it was introduced). As an example:

image

This can be demonstrated in v1 (`develop`) using the ENG keyboard and trying to type `@` (which is Shift-D2):
image

Since you use POR, you do not see this because you coded all this to assume POR. With the POR keyboard:
image

Likewise, `ConsoleKeyMappingTests.GetConsoleKeyInfoFromKeyCode_Tests` is (and has always been) broken:

image

This is broken too:

image

In short, your understanding of how VK_PACKET works is incorrect. Previously, I was confused too (in a different way). 

I am currently working on this and will submit a PR to PR with my suggested changes in the next few days. Please work on something else in the meantime so we are not working against each other.

**Comment by BDisp** [comment]
> @BDisp please read this: https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-mapvirtualkeya
> 
> Esp:

image

This was the first link where I learn more about VK and VSC. Yes is true virtual keys codes from A..Z are translated to upper-case A..Z characters regardless of current keyboard layout. What it's saying here is that the `ConsoleKey` are always A..Z no matter the unicode character is a..z, à..ý. If you want to translate a virtual-key code to the corresponding character, use the ToAscii function, which will check the uVirtKey, uScanCode and lpKeyState. So if it's `VK_A, 30` the result is `a`. If it's `VK_A, 30, shift` the result will be `A`.

> This can be demonstrated in v1 (develop) using the ENG keyboard and trying to type @ (which is Shift-D2):

In the v1 is still more buggy than the v2. In v2 I fixed more bugs. In PTG keyboard I have two options to get the `@`, by type `Alt Gr+D2` or `Ctrl+Alt+D2`. The `Alt Gr` is equivalent to press `Ctrl+Alt`.

> Since you use POR, you do not see this because you coded all this to assume POR. With the POR keyboard:

Yes is true, I know this was broken other keyboard layouts. There are many code pages and perhaps adding one more property in the `ScanCodeMapping` with the corresponding right character for each code page.

> Likewise, ConsoleKeyMappingTests.GetConsoleKeyInfoFromKeyCode_Tests is (and has always been) broken:

Yes as I said before, depend of the code page that a keyboard is using.

> In short, your understanding of how VK_PACKET works is incorrect. Previously, I was confused too (in a different way).

Well if you talking about how different layout keyboards interpret it I understand. But I admit my PR isn't totally correct as I think you also are not totally correct. I hope you will submit something that we can find a common agreement.

**Comment by tig** [comment]
> Using a virtual keyboard isn't the same as a real keyboard.

I am NOT using the virtual keyboard. I am using my real keyboard, but changing the keyboard layout. 

Thing's I've learned:

In VS Code, see the "Replace with Nex/Previous Value" key bindings ("Ctrl+Shift+." and "Ctrl+Shift+,") as an example.  

I am using my real keyboard with the ENG, POR, or DEU keyboard mappings enabled. For example, if I select POR:

image

I need to press the OemPeriod key to get `.` and Shift-OemPeriod to get `:` (on the ENG keyboard OemPeriod is `.`/`>`). 

OemPeriod is a good test case for this because VK_OEM_PERIOD is documented as: "For any country/region, the '.' key". Note it says nothing about `>` because that is US (ENG) specific. There IS NO `>` key on ANY keyboard. There IS an Oem102 key though:

On the POR keyboard, `>` is `Shift+Oem102`:

```
event: [KeyEventRecord(up,1,226,86,>,ShiftPressed)] [ConsoleKeyInfoEx(Key: 226 (Oem102) | Shift, KeyChar: > (62))]
```

The correct processing for all this is to use the Win32 API "MapVirutalKey" with the `MAPVK_VK_TO_CHAR` setting:

https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-mapvirtualkeyexa

Something else I've learned: If the keyboard layout changes Windows Terminal (or whatever is processing all this) is NOT listening to the `WM_INPUTLANGCHANGE` message (https://learn.microsoft.com/en-us/windows/win32/winmsg/wm-inputlangchange). This means that `MapVirtualKey` always maps based on whatever the keyboard layout was when the process started. 

This can be reproduced at the command line by:

1) Set the Keyboard layout to ENG (the key two keys to the left from Enter will be `;`/`:` which is VK_OEM_1).
2) Start a WT session
3) Press the `;`/`:` key
image

4) Press `Shift-;`:
image

5) Press `Ctrl-;` (WT strips ctrl off any Ctrl key that is not bound to an action):
image

6) Switch the keyboard layout to POR (Win-Space cycles through loaded layouts).

7) Press the `VK_OEM_1` key (on POR this is the key labeled `ç`. As expected:

image

8) Press the `Shift-VK_OEM_1` key. As expected:

image

9) Now, this is where it gets interesting. Press `Ctrl-VK_OEM_1`. This SHOULD print `ç`...

image

10) To prove this, start a fresh WT console and Press `Ctrl-VK_OEM_1`

image

Note, you can also cause the new keyboard layout to be noticed by simply starting a new `pwsh` instance.

The upshot of this is: Be careful in your testing ... changing keyboard layouts while a Terminal Gui app is running will give confusing results!

I've tried using `GetKeyboardLayout` and it always reports the same layout that was active when the process started. 

I plan on submitting a WT bug for this.

I also plan on trying to find a work around because it's super annoying!

**Comment by BDisp** [comment]
> Here's the WT Bug: https://github.com/microsoft/terminal/issues/8871

Good catch.

> I also plan on trying to find a work around because it's super annoying!

Yes, I'm expecting that.

In the v1 when I started the VK Packet Simulator I used the Win32 API but I had problems when I run the scenario in Linux because Win32 API isn't supported. I end up by creating the class to map virtual keys, unicode char and modifiers to a `ConsoleKeyInfo`. The scan codes is only needed for `WindowsDriver`. But I think the `WindowsDriver` is a good reference to learn more hw about this work. I know you switch from ENG to POR with the VK_OEM_1 but can you confirm on `WindowsDriver` if the `Key` appear as `Oem1` or `Oem3` on POR layout, please.

When I press `Ctrl+ç` I get this values on the breakpoint. See that the `KeyChar` is `'\0'` and `Modifiers` is `Control` which normally means that it won't be a printable character but only a key binding.

Captura de ecrã 2024-01-01 182234

When I press `Ctrl+Shift+ç` I get this values on the breakpoint. See that the `KeyChar` is `'\0'` and `Modifiers` is `Shift | Control` which normally means that it won't be a printable character but only a key binding.

Captura de ecrã 2024-01-01 182446

When I press `ç` I get this values on the breakpoint. See that the `KeyChar` is `'ç'` and `Modifiers` is `None` which normally means that it is a printable character.

Captura de ecrã 2024-01-01 182531

When I press `Shift+ç` I get this values on the breakpoint. See that the `KeyChar` is `'Ç'` and `Modifiers` is `Shift` which normally means that it is a printable character. But remember that we need to remove the shift before send the key to `TextView` or other view that only must process the `KeyChar`.

Captura de ecrã 2024-01-01 182608

When I press `Ctrl+Alt+ç` I get this values on the breakpoint. See that the `KeyChar` is `'\0'` and `Modifiers` is `Alt | Control` which normally means that it won't be a printable character but only a key binding. There are case where using `Ctrl+Alt+KeyChar` or `Alt gr+KeyChar` is a valid printable character on POR layout, like the @, £, §, €, {, [, ], } where the `KeyChar` will be different of `'\0'`.

Captura de ecrã 2024-01-01 182643

When I press `Shift+Alt+ç` I get this values on the breakpoint. See that the `KeyChar` is `'Ç'` and `Modifiers` is `Alt | Shift` which normally means that it won't be a printable character but only a key binding, although `KeyChar` is `'Ç'`.

Captura de ecrã 2024-01-01 182716

**Comment by tig** [comment]
I'll push a PR soon so you can test yourself.

Does anyone know why `KeyCode` (originally `Key`) used seemingly random values for the members above `CursorUp`?

Why not just have these be `ConsoleKey.CursorUp + 0x100000`?

If this where the case, mapping would be SO MUCH simpler in both WindowsDriver and NetDriver!

**Comment by BDisp** [comment]
> I'll push a PR soon so you can test yourself.
> 
Nice.

> Does anyone know why `KeyCode` (originally `Key`) used seemingly random values for the members above `CursorUp`?
> 
The ConsoleKey.UpArrow = 38 which represent the ASCII `&`. Thus `KeyCode.CursorUp` cannot have that value. Was necessary to set a value greater than the last valid printable character. So all the members above the `CursorUp` also have a value increased by one. We know that a enum flags should be set as 1, 2, 4, 8, 16, 32, etc. but the intention was set the printable values as ASCII values and the equivalent to the `ConsoleKey` as mask with high values.

> Why not just have these be `ConsoleKey.CursorUp + 0x100000`?
> 
I think so, if they don't overflow.

> If this where the case, mapping would be SO MUCH simpler in both WindowsDriver and NetDriver!

I think they are very simple to mapping to a `ConsoleKeyInfo`. The problem with `NetDriver` is the `Shift` key doesn't come in the modifiers variable because t's using a virtual terminal which also is having the same behavior as Linux have.

**Comment by tig** [comment]
What da fuq is the difference between `KeyCode.Delete` and `KeyCode.DeleteChar`?

**Comment by BDisp** [comment]
This is bad in my opinion. Currently `KeyCode.Delete=127` which represent the ASCII 127 (DEL) and the `KeyCode.DeleteChar` is a masked value that really represent the `ConsoleKey.Delete=46`, which represent the ASCII `.` and thus must be a masked value. In my opinion we should remove the `KeyCode.Delete` and replace the `KeyCode.DeleteChar` with the `KeyCode.Delete`, because we don't need the ASCII 127 represented as `KeyCode`, I think.
And the `KeyCode.InsertChar` should be replaced with `KeyCode.Insert` because `ConsoleKey.Insert=45`, which represent the ASCII `-` and we only must use the masked value for `KeyCode.Insert`. Do you agree?

**Comment by dodexahedron** [comment]
I think @BDisp is encountering what I was ultimately getting at but not communicating properly. 😅 

Isn't globalization fun?

**Comment by tig** [comment]
@BDisp do you know what this is?

image

**Comment by BDisp** [comment]
> @BDisp do you know what this is?
> 
> image

I think I used that a long time ago to use on the `Application` class to get some request, like the terminal type, but it mustn't work like that. Can be removed.
